crates/bitwarden/src/tool/generators/password.rs

```diff
@@ -40,10 +41,249 @@ pub struct PassphraseGeneratorRequest {
+        );
+        let special = chars(special, SPECIAL_CHARS, &[], avoid_ambiguous);
+        let all = lower
+            .iter()
+            .chain(&upper)
+            .chain(&number)
+            .chain(&special)
+            .copied()
+            .collect();
+
+        Self {
+            lower,
+            upper,
+            number,
+            special,
+            all,
+        }
+    }
+}
+
+pub(super) fn password(input: PasswordGeneratorRequest) -> Result<String> {
+    password_with_rng(rand::thread_rng(), input)
+}
+
+pub(super) fn password_with_rng(
+    mut rng: impl RngCore,
+    input: PasswordGeneratorRequest,
+) -> Result<String> {
+    // We always have to have at least one character set enabled
+    if !input.lowercase && !input.uppercase && !input.numbers && !input.special {
+        return Err(Error::Internal(
+            "At least one character set must be enabled",
+        ));
+    }
+
+    // Generate all character dictionaries
+    let chars = PasswordGeneratorCharSet::new(
+        input.lowercase,
+        input.uppercase,
+        input.numbers,
+        input.special,
+        input.avoid_ambiguous.unwrap_or(false),
+    );
+
+    // Make sure the minimum values are zero when the character
+    // set is disabled, and at least one when it's enabled
+    fn get_minimum(min: Option<u8>, enabled: bool) -> u8 {
+        if enabled {
+            u8::max(min.unwrap_or(1), 1)
+        } else {
+            0
+        }
+    }
+    let min_lowercase = get_minimum(input.min_lowercase, input.lowercase);
+    let min_uppercase = get_minimum(input.min_uppercase, input.uppercase);
+    let min_number = get_minimum(input.min_number, input.numbers);
+    let min_special = get_minimum(input.min_special, input.special);
+
+    // Check that the minimum lengths aren't larger than the password length
+    let min_length = min_lowercase + min_uppercase + min_number + min_special;
+    let length = input.length.unwrap_or(DEFAULT_PASSWORD_LENGTH);
+    if min_length > length {
+        return Err(Error::Internal(
+            "Password length can't be less than the sum of the minimums",
+        ));
+    }
+
+    // Generate the minimum chars of each type, then generate the rest to fill the expected length
+    let mut buf = Vec::with_capacity(length as usize);
+
+    for _ in 0..min_lowercase {
+        buf.push(*chars.lower.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in 0..min_uppercase {
+        buf.push(*chars.upper.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in 0..min_number {
+        buf.push(*chars.number.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in 0..min_special {
+        buf.push(*chars.special.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in min_length..length {
+        buf.push(*chars.all.choose(&mut rng).expect("slice is not empty"));
+    }
+
+    buf.shuffle(&mut rng);
+    Ok(buf.iter().collect())
```

Comment:
📝  Here's where you use `Distribtion<char>.sample_iter`. Create each iterator, then take the number you need. Chain the takes together, and collect them in a Vec. Shuffle the vec, then convert to a string.

:+1: Put another way, this code is doing exactly what it should, given that you're working with arrays.